crud/insert_many.lua

```diff
@@ -0,0 +1,256 @@
+local checks = require('checks')
+local errors = require('errors')
+local vshard = require('vshard')
+
+local call = require('crud.common.call')
+local const = require('crud.common.const')
+local utils = require('crud.common.utils')
+local sharding = require('crud.common.sharding')
+local dev_checks = require('crud.common.dev_checks')
+local schema = require('crud.common.schema')
+
+local BatchInsertIterator = require('crud.common.map_call_cases.batch_insert_iter')
+local BatchPostprocessor = require('crud.common.map_call_cases.batch_postprocessor')
+
+local InsertManyError = errors.new_class('InsertManyError', {capture_stack = false})
+
+local insert_many = {}
+
+local INSERT_MANY_FUNC_NAME = '_crud.insert_many_on_storage'
+
+local function insert_many_on_storage(space_name, tuples, opts)
+    dev_checks('string', 'table', {
+        add_space_schema_hash = '?boolean',
+        fields = '?table',
+        stop_on_error = '?boolean',
+        rollback_on_error = '?boolean',
+        sharding_key_hash = '?number',
+        sharding_func_hash = '?number',
+        skip_sharding_hash_check = '?boolean',
+    })
+
+    opts = opts or {}
+
+    local space = box.space[space_name]
+    if space == nil then
+        return nil, {InsertManyError:new("Space %q doesn't exist", space_name)}
+    end
+
+    local _, err = sharding.check_sharding_hash(space_name,
+                                                opts.sharding_func_hash,
+                                                opts.sharding_key_hash,
+                                                opts.skip_sharding_hash_check)
+
+    if err ~= nil then
+        return nil, {err}
+    end
+
+    local inserted_tuples = {}
+    local errs = {}
+
+    box.begin()
+    for _, tuple in ipairs(tuples) do
+        -- add_space_schema_hash is true only in case of insert_object_many
+        -- the only one case when reloading schema can avoid insert error
+        -- is flattening object on router
+        local insert_result = schema.wrap_box_space_func_result(space, 'insert', {tuple}, {
+            add_space_schema_hash = opts.add_space_schema_hash,
+            field_names = opts.fields,
+        })
+
+        if insert_result.err ~= nil then
+            local err = {
+                err = insert_result.err,
+                tuple = tuple,
+            }
+
+            if opts.stop_on_error == true then
+                if opts.rollback_on_error == true then
+                    box.rollback()
+                    return nil, {err}
+                end
+
+                box.commit()
+
+                return inserted_tuples, {err}
+            end
+
+            table.insert(errs, err)
+        end
+
+        table.insert(inserted_tuples, insert_result.res)
+    end
+
+    if next(errs) ~= nil then
+        if opts.rollback_on_error == true then
+            box.rollback()
+            return nil, errs
+        end
```

Comment:
We should add all non-inserted tuples into errors, including ones that are rolled back. Otherwise a user will have no idea what is actually inserted and what is not inserted.